### ISSUE TITLE
Difference between model vs model_seq

### ISSUE THREAD

**Issue by iMaTzzz** [MIV-XJTU/ARTrack#110]
What does seq mean in the code ? (e.g. artrackv2_seq)

And what's the difference between artrackv2 vs artrackv2_seq ? Which one should we use ?

Thanks!

**Comment by AlexDotHam** [comment]
artrackv2 is for the one-stage training, per-frame template matching tracking pipeline, for speed up the training processing of sequential training. artrackv2_seq same as artrack_seq, for the second-stage training and inference, which is sequential level tracking pipeline.

**Comment by iMaTzzz** [comment]
Thank you for your answer.

I tried to make some test using ARTrackV2 and the speed upgrade is barely noticeable compared to ARTrack. Do I need to specify a parameter ? From my own tests, I had a 50% speed up (1.5x), is it normal ? I used these weights: ARTrackV2-B-256

**Comment by AlexDotHam** [comment]
> Thank you for your amazing work. The checkpoints you release seems to be artrackv2? Not for artrackv2-seq? At least when I try the ARTrackV2_large_384_got.pth.tar from Baidu Yunpan, it seems to be artrackv2?

This is just the wrong name, the checkpoints is for ARTrackV2-seq, however when I upload the checkpoints for clarify which is belong to large, I rename it. Sorry for misunderstanding.

**Comment by AlexDotHam** [comment]
> Thank you for your answer.
> 
> I tried to make some test using ARTrackV2 and the speed upgrade is barely noticeable compared to ARTrack. Do I need to specify a parameter ? From my own tests, I had a 50% speed up (1.5x), is it normal ? I used these weights: ARTrackV2-B-256

This may be because of the influence of Reconstruction Decoder, this part needs to be designed because the masked attention is slower than standard attention mechanism, beyond that the feature transport between search into reconstruct decoder is slow. Moreover, during inference, the attention in vit should be replaced with the scaled_dotsproduct_attention in torch.nn.functrion, which helps to use GPU-compatible attention such as Flash Attn
Use contiguous to ensure memory consistency and continuity when transmitting images feature output to the decoder.

**Comment by iMaTzzz** [comment]
> This may be because of the influence of Reconstruction Decoder, this part needs to be designed because the masked attention is slower than standard attention mechanism, beyond that the feature transport between search into reconstruct decoder is slow. Moreover, during inference, the attention in vit should be replaced with the scaled_dotsproduct_attention in torch.nn.functrion, which helps to use GPU-compatible attention such as Flash Attn
Use contiguous to ensure memory consistency and continuity when transmitting images feature output to the decoder.

Thanks a lot, I will try to implement it and so some tests 👍 

Also, do you now have a recommendation for a function to detect when the model has lost track of the object ?
In ARTrack, you recommended using the predicted sequence scores from the AI output but it wasn't super precise. But now, in ARTrackV2, there is also the predicted score. So I came up with a function which computes the average of the last predicted scores and check how far away it is from the current predicted score. Do you have any suggestions ?

**Comment by AlexDotHam** [comment]
> > This may be because of the influence of Reconstruction Decoder, this part needs to be designed because the masked attention is slower than standard attention mechanism, beyond that the feature transport between search into reconstruct decoder is slow. Moreover, during inference, the attention in vit should be replaced with the scaled_dotsproduct_attention in torch.nn.functrion, which helps to use GPU-compatible attention such as Flash Attn
> > Use contiguous to ensure memory consistency and continuity when transmitting images feature output to the decoder.
> 
> Thanks a lot, I will try to implement it and do some tests 👍
> 
> Also, do you now have a recommendation for a function to detect when the model has lost track of the object ? In ARTrack, you recommended using the predicted sequence scores from the AI output but it wasn't super precise. But now, in ARTrackV2, there is also the predicted score. So I came up with a function which computes the average of the last predicted scores and check how far away it is from the current predicted score. Do you have any suggestions ?

Essentially, I don't think the model can predict whether it has lost its target. It's like saying that if the model knows it has lost its target, why doesn't it pursue other targets? So there must be a delay in the score of this tracking, for example, when I lose multiple frames, the model will only realize that the target has disappeared. Therefore, directly calculating the L2 distance between the features of the dynamic template and the original template is a reasonable solution in my opinion, rather than using the model to generate the score itself.
